Add Gps Sensor to IGN-Sensors library

Dotrar opened the pull request:
Related issues: 
https://github.com/ignitionrobotics/ign-sensors/issues/23
https://github.com/osrf/sdformat/pull/453

very similar to the imu and altimeter sensor objects, just a gps sensor, hopefully as one would expect.

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2019 Open Source Robotics Foundation
```
2020

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+      /// \param[in] _now The current time
+      /// \return true if the update was successfull
+      public: virtual bool Update(
+        const std::chrono::steady_clock::duration &_now) override;
+
+
+
+      /// \brief Set the latitude of the GPS
```
```suggestion
        const std::chrono::steady_clock::duration &_now) override;

      /// \brief Set the latitude of the GPS
```

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+
+      /// \brief Get the latitude of the GPS, wraped between +/- 180
+      /// \return Latitude in degrees.
+      public: double Latitude() const;
+
+
+      /// \brief Set the longitude of the GPS
```
```suggestion
      public: double Latitude() const;

      /// \brief Set the longitude of the GPS
```

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+
+      /// \brief Get the longitude of the GPS, wraped between +/- 180
+      /// \return Longitude in degrees.
+      public: double Longitude() const;
+
+
+      /// \brief Set the altitude of the GPS
```
```suggestion
      public: double Longitude() const;

      /// \brief Set the altitude of the GPS
```

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+
+      /// \brief GPS altitude is the GEOMETRIC altitude above sea level
+      /// \return Altitude in meters
+      public: double Altitude() const;
+
+
+      /// \brief Set the velocity of the GPS
```
```suggestion
      public: double Altitude() const;

      /// \brief Set the velocity of the GPS
```

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+
+      /// \brief Get the velocity of the GPS sensor
+      /// \return velocity vector3 in meters per second
+      public: ignition::math::Vector3d Velocity() const;
+
+
+      /// \brief Easy short hand for setting the lat/long of the gps.
```
```suggestion
      public: ignition::math::Vector3d Velocity() const;

      /// \brief Easy short hand for setting the lat/long of the gps.
```

ahcorde commented on include/ignition/sensors/GpsSensor.hh:
```diff
@@ -0,0 +1,134 @@
+      /// \brief Easy short hand for setting the lat/long of the gps.
+      /// \param[in] _latitude in degrees
+      /// \param[in] _longitude in degrees
+      public: void SetPosition(double _latitude, double _longitude, double _altitude = 0.0);
+
+
+      IGN_COMMON_WARN_IGNORE__DLL_INTERFACE_MISSING
```
```suggestion
      public: void SetPosition(double _latitude, double _longitude, double _altitude = 0.0);

      IGN_COMMON_WARN_IGNORE__DLL_INTERFACE_MISSING
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (C) 2019 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2020 Open Source Robotics Foundation
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  // Apply gps position noise
+  if (this->dataPtr->noises.find(GPS_POSITION_NOISE) !=
+      this->dataPtr->noises.end())
+  {
+
+    // Apply in degrees. 
```
```suggestion
  {
    // Apply in degrees. 
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+
+  //normalise so that it is within +/- 180
+  this->dataPtr->latitude.Normalize();
+  this->dataPtr->longitude.Normalize();
+
+
+  msg.set_latitude_deg(this->dataPtr->latitude.Degree());
```
```suggestion
  this->dataPtr->longitude.Normalize();

  msg.set_latitude_deg(this->dataPtr->latitude.Degree());
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  this->dataPtr->latitude.Degree(_latitude);
+}
+
+double GpsSensor::Latitude() const {
+  return this->dataPtr->latitude.Degree();
```
```suggestion
double GpsSensor::Latitude() const
{
  return this->dataPtr->latitude.Degree();
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  return this->dataPtr->latitude.Degree();
+}
+
+void   GpsSensor::SetAltitude(double _altitude) {
+  this->dataPtr->altitude = _altitude;
```
```suggestion
void GpsSensor::SetAltitude(double _altitude)
{
  this->dataPtr->altitude = _altitude;
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  return this->dataPtr->altitude;
+}
+
+void   GpsSensor::SetLongitude(double _longitude) {
+  this->dataPtr->longitude.Degree(_longitude);
```
```suggestion
void GpsSensor::SetLongitude(double _longitude)
{
  this->dataPtr->longitude.Degree(_longitude);
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  this->dataPtr->longitude.Degree(_longitude);
+}
+
+double GpsSensor::Longitude() const {
+  return this->dataPtr->longitude.Degree();
+}
```
```suggestion
double GpsSensor::Longitude() const
{
  return this->dataPtr->longitude.Degree();
}
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  return this->dataPtr->longitude.Degree();
+}
+
+void   GpsSensor::SetVelocity(ignition::math::Vector3d &_vel) {
+  this->dataPtr->velocity = _vel;
+}
```
```suggestion
void GpsSensor::SetVelocity(ignition::math::Vector3d &_vel)
{
  this->dataPtr->velocity = _vel;
}
```

ahcorde commented on src/GpsSensor.cc:
```diff
@@ -0,0 +1,246 @@
+  return this->dataPtr->velocity;
+}
+
+void   GpsSensor::SetPosition(double _latitude, double _longitude, double _altitude) {
+  this->SetLongitude(_longitude);
```
```suggestion
void GpsSensor::SetPosition(double _latitude, double _longitude, double _altitude)
{
  this->SetLongitude(_longitude);
```

ahcorde reviewed (changes requested):
Please review the style comments and check the CI you have some linters errors.

Do you mind to add some tests too ? (like these ones https://github.com/ignitionrobotics/ign-sensors/tree/ign-sensors4/test/integration)

Dotrar commented:
Added some tests and I fixed up the linting errors from the ./tools/code_check.sh script. 

this is going to need https://github.com/osrf/sdformat/pull/453 to be accepted before it willl compile here, just for the fyi.

I think most of the help I might need would be on the gazebo sim side. https://github.com/ignitionrobotics/ign-gazebo/pull/519 

I'll fix up the lint on there soon.

chapulina commented:
@Dotrar , the SDFormat PR has been merged. Are you interested in iterating on this PR and the one in `ign-gazebo`?

chapulina commented:
Thank you again for all the work, @Dotrar . Since the fork has been deleted, I opened a new PR in #177.
